g:clever_f_mark_direct extended beyond ASCII

jidhub opened the pull request:
`ch`, instead of just containing a byte (that is ignored when non-ASCII, e.g., coding utf8 or other encoding), now contains a possibly multibyte char. `line` is changed to contain the list of these groupped non-ASCII chars. Tested on utf8.

jidhub committed:
g:clever_f_mark_direct extended beyond ASCII

`ch`, instead of just containing a byte (that is ignored when non-ASCII, e.g., coding utf8 or other encoding), now contains a possibly multibyte char.
`line` is changed to contain the list of these groupped non-ASCII chars.
Tested on utf8.

rhysd reviewed (changes requested):
Could you add a test case which tests multi-byte characters to this test suite?

https://github.com/rhysd/clever-f.vim/blob/e852984d2bb6671034851fa70442d7799c1d7856/test/test.vimspec#L1223

rhysd commented:
Could you ensure all tests passed on your local at first? Here is a guide to run tests:

https://github.com/rhysd/clever-f.vim/blob/master/test/README.md

jidhub committed:
Update autoload/clever_f.vim thanking rhysd

Co-authored-by: Linda_pp <rhysd@users.noreply.github.com>

jidhub committed:
Update autoload/clever_f.vim thanking rhysd

Co-authored-by: Linda_pp <rhysd@users.noreply.github.com>

jidhub commented:
Sure will do.

> Could you ensure all tests passed on your local at first? Here is a guide to run tests:

The test that fails is the test checking that clever_f#_mark_direct() only highlight ASCII chars : "not ok 65 - clever_f#_mark_direct() should not highlight multibyte characters"

Can I update this test, instead of creating a new test ?

jidhub committed:
Update autoload/clever_f.vim thanks rhysd

Co-authored-by: Linda_pp <rhysd@users.noreply.github.com>

jidhub commented:
There is also a spurious vint[1] error: 
      `autoload/clever_f.vim:160:29: Undefined variable: c: (see :help E738)`

`c` is defined at line 146: `let [_, l, c, _] = getpos('.')`

and used just above my patch at line 148: `if (a:forward && c >= len(line)) || (!a:forward && c == 1)`

How comes that vint suggests `c` has been undefined ?

--
[1] https://github.com/rhysd/clever-f.vim/pull/74?notification_referrer_id=NT_kwDOAIcHSrI0NTk5NTM3MDA2Ojg4NDkyMjY

jidhub committed:
patching multibyte test for clever_f#mark_direct

jidhub committed:
updating description of the test

jidhub commented:
Oh sorry I use github editor and was prepared to run tests locally afterwards.

jidhub committed:
don't return last char of multibyte under cursor 

CI spotted that bug. I did not yet run CI on my computer (I test on vim in termux app of google play).

jidhub commented:
Githud says "Changes requested" but I am not finished testing !

jidhub committed:
Fixed test, added backwards test

I don't know why getmatches() is 1-based and not 0-based

jidhub commented:
I added a backwards test. Now I need to install the testing software on my computer to get some more input.

jidhub committed:
all test pass with c-1 in line 161.

# tests 81
# passes 81

jidhub committed:
skip start of current char getline('.')[c-1] for F

After `let [_, l, c, _] = getpos('.')`, first byte of char under cursor is `getline('.')[c-1]`. Hence don't include, and don't skip it, when going backwards.
Also, mark_direct has always been slow on very long lines. I commented about optimization, perfect for `f`, and imperfect for `F`.

jidhub committed:
updated comments, please migrate them to an issue

I made better comments. After the pull request passed, I will migrate these comments to an issue (I don't know how to add a new option)

jidhub committed:
range should stay positive in the future issue

jidhub committed:
Update clever_f.vim

jidhub committed:
Update clever_f.vim

jidhub commented:
Test passed, this is good to go !
/data/data/com.termux/files/home/.vim/pack/bundle/start/clever-f.vim-mark_direct_beyond_ASCII/test$ ./vim-themis/bin/themis  .
# tests 81
# passes 81

Thanks for your review time and insights.

jidhub commented:
Github still repeats your request to create a new multibyte tescase: https://github.com/rhysd/clever-f.vim/pull/74#pullrequestreview-1141711381

How can I tell github that I did exactly that in https://github.com/rhysd/clever-f.vim/pull/74/commits/ddf8b51403e34b83d2cd04585f33b1def140297b ?

rhysd commented:
> Github still repeats your request to create a new multibyte tescase: https://github.com/rhysd/clever-f.vim/pull/74#pullrequestreview-1141711381

That's because I haven't reviewed this PR again. What you need to do is just requesting me review again. Don't mind, I'll review this PR soon.

rhysd commented:
Looks good. Thank you for adding tests as well.

rhysd commented on autoload/clever_f.vim:
```diff
@@ -156,11 +156,18 @@ function! clever_f#_mark_direct(forward, count) abort
-        " only matches to ASCII
-        if ch !~# '^[\x00-\x7F]$' | continue | endif
+    if a:forward 
+        let line = split(line[c - 1 : ], '\zs') " adding a limit like c+4*termcol*termlines just after `:` is a free gift in a very long line, assuming all char are less than 4 bytes.
```
Note that this does not work since some characters are longer than 4 bytes like 👪 or 🇺🇸

jidhub replied:
Agreed, lines can wrap. Hence the screen width is not enough, that is why I multiplied the number of columns of the screen by the number of lines in the screen.

jidhub commented:
Thanks for the additionnal reformatting.

Can you tel me if the following coverage notice is positive or negative ?

     ##           master      #74      +/-   ##
    ==========================================
    + Coverage   93.50%   93.64%   +0.14%     
    ==========================================
      Files           2        2              
      Lines         354      362       +8     
    ==========================================
    + Hits          331      339       +8     
      Misses         23       23              

Can you share a link about what coverage does ?

rhysd commented:
It means positive (+8 lines)
